### ISSUE TITLE
Add ability to reset state of Apollo mutation

### ISSUE THREAD

**Issue by avasuro** [apollographql/apollo-feature-requests#170]
# The problem
There is no ability to reset mutation state to initial (when mutation was not invoked)

# Solution
I propose to return "reset" function as third item of array, returning from useMutation hook, like this:
```javascript
const [mutate, mutationResult, resetMutation] = useMutation(SOME_MUTATION);
```

# Motivation
It will be very convenient to have ability to reset mutation state to initial in some cases.
For example: we have a login page, when user clicks on "login" if login failed - show modal window with error message and button "OK" which should close this window. This can be achieved using the following code:

```javascript
function LoginPage () {
    const [
        login,
        { error, data, called }
    ] = useMutation(LOGIN_MUTATION);

    return (
        <>
            <form>
                <input class="login"/>
                <input class="password"/>
                <button onclick={login}>Login</button>
            </form>
            {
                error && 
                <LoginFailedMessageWindow
                    message={error}
                    onOkClicked={() => {
                        // What we can write here to close the window (to reset mutation state)?
                    }}
                />
            }
        </>
    )
}
```

But the problem is that after user clicked on "ok" button in LoginFailedMessageWindow - there is no ability to reset mutation state to initial, so to implement window closing I forced to implement my own mutation state management.

<details><summary> expend this section if you are interested in implementation details of mutation state management hook</summary>
<p>

```javascript
import {useState, useCallback} from 'react';

const MUTATION_STATES = {
    IDLE: 'idle',
    SUCCEED: 'succeed',
    FAILED: 'failed'
};

const INITIAL_STATE = {
    state: MUTATION_STATES.IDLE,
    error: null
};

function useMutationState(mutationFunc) {
    const [mutationState, setMutationState] = useState(INITIAL_STATE);
    const wrappedMutation = useCallback((...args) => {
        const mutation = mutationFunc(...args);
        mutation.then(
            (result) => {
                console.log(result);
                setMutationState({
                    state: MUTATION_STATES.SUCCEED,
                    data: result,
                    error: null
                });
            },
            (e) => {
                setMutationState({
                    state: MUTATION_STATES.FAILED,
                    error: e
                });
            }
        );
        return mutation;
    }, mutationFunc);
    const resetMutationState = useCallback(() => {
        setMutationState(INITIAL_STATE);
    }, []);
    return [
        mutationState,
        wrappedMutation,
        resetMutationState
    ];
}

export {
    MUTATION_STATES,
    useMutationState
};
```

</p>
</details>

and use it like that:

```javascript
function LoginPage () {
    const [loginMutation] = useMutation(LOGIN_MUTATION);
    const [loginState, login, resetLoginState] = useMutationState(loginMutation);

    return (
        <>
            <form>
                <input class="login"/>
                <input class="password"/>
                <button onclick={login}>Login</button>
            </form>
            {
                loginState.error && 
                <LoginFailedMessageWindow
                    message={error}
                    onOkClicked={resetLoginState}
                />
            }
        </>
    )
}
```

So, in particular I implement the same logic that already implemented in apollo mutation result (and pay attention that here I don't handle states like "loading" "called" etc, if I will implement them the source code of mutation state management hook will be even more complicated)

**Comment by nratter** [comment]
Any ideas on how to accomplish this with Apollo 3?

**Comment by avasuro** [comment]
Here is a possible implementation. There is just one question - what to do with pending mutations? I don't know whether there is some way to abort pending mutation. I need help here.

**Comment by danieldunderfelt** [comment]
Either I'm missing something about how to use Mutation hooks or this should have been one of the first features implemented for it. How is the mutation hook supposed to be used without wiring up separate state management for both the data and the errors?

If I need to call the mutation multiple times in the same component, a way to reset it is absolutely required in order to provide a nice UI. Otherwise I need separate state to track what data or errors were returned from which mutation activation.

**Comment by Samuelote** [comment]
I'm seconding what @danieldunderfelt said. Any great way of doing this ATM?

**Comment by avasuro** [comment]
@galmail
I'm not quite sure how PR that you've mentioned correlates with this ticket (except same type of issue).
This PR is for `graphql-hook` package repo, and If I understand properly Apollo don't uses this package under the hood - it uses own queries management class called "QueryManager", which uses "MutationStore" class under the hood  to store mutations results, and "ApolloLink" class to resolve requests.

And as @danieldunderfelt said - in theory functionality for reset state/request should have been implemented initially, but in reality is it is not.

Just to clarify, I will quickly explain how Apollo "useMutation" method works now, then will describe the problem with resetMutation that I see in details.

### useMutation internals

When user calls `useMutation()` hook - source code of this hook calls:
`src/react/data/MutationData.ts -> execute()` which returns `[runMutation, mutationResults]` (this is the array which then returned from `useMutation` hook)

After user calls `runMutation` the following sequence of calls is executed:
`src/react/data/MutationData.ts -> mutate()` which calls:
`src/ApolloClient.ts -> mutate()` which calls:
`src/core/QueryManager.ts -> mutate()` which under the hood does a lot of things, but most important are:
- generate new mutation ID
```
const mutationId = this.generateQueryId();
```
- create initial state for new mutation in mutation store:
```
this.mutationStore.initMutation(
  mutationId,
  mutation,
  variables,
);
```
- create and return new Promise which sends mutation request to server (source code of this part blown my mind :) ). Under the hood the following methods are called:
`src/core/QueryManager.ts -> getObservableFromLink()` which calls
`src/link/core/ApolloLink.ts -> execute()` which calls
`src/link/core/ApolloLink.ts -> request()`

### The problem
New method "resetMutation" can work good in all cases, except when this method will be called on pending mutation. And the question is - how it should behave in that case? I see two possible ways:
1) Do nothing (as it is implemented now)
2) Abort pending mutation and reset mutation state to initial 

Second way sounds good, but there is the following difficulties:
- How to stop pending request?
- How to abort promise returned from runMutation function?

#### How to stop pending request?
As seen from useMutation internals - Apollo doesn't stick to some specific transport protocol implementation (like HTTP or WebSocket), and uses common interface of ApolloLink class to manage requests, and this interface doesn't have any method to stop pending requests. So to implement this - some method should be added to ApolloLink class. There is may be just one caveat - probably some transport protocols may not support abort functionality, I don't know how it's better to manage that cases, may be abort method of ApolloLink should be "optional" (like return true/false if success/fail or throw an error, so Apollo source code will always know whether abort was successful or not)

#### How to abort promise returned from runMutation function?
important: In that case "abort" promise should not be equal to "reject" promise (at least after promise is rejected - rejection error should not appear in "error" property of mutationResults object, which returned from useMutation hook), so in promise error handler we should distinguish between rejection because promise was aborted, and rejection because any other reason. There is `browser-cancellable-events` package which provides ability to abort promises (as well as many other async operations), and I like [approach that they use](https://www.npmjs.com/package/browser-cancelable-events#promise-example) - in promise rejection handler they check whether rejection error is not an instance of CancelledPromiseError class.

Back to our QueryManager class: currently there is no way to abort pending mutation promise - there is no method to reject some specific pending mutation. If you look at the `src/core/QueryManager.ts -> query()` function - it uses `fetchQueryRejectFns` private property to collect all functions to reject pending queries (and then this property used in `src/core/QueryManager.ts -> clearStore()` method), I think something similar can be used to reject pending mutations.

**Comment by tomerzcod7** [comment]
Any news on this? I'm facing the same problem. Using useMutation in my form, and using the error variable to show messages on the UI. After showing the message, if the user changes one of the inputs I want the error to go away when the state of the component changes. However, the data and error variables persist.

**Comment by raulingg** [comment]
any news ?

**Comment by markymc** [comment]
Just discovered I need this too. Seems super important!

**Comment by marc-ed-raffalli** [comment]
This feature would have been handy indeed. Implemented a quick wrapper to work around the limitation for now.
Sharing below, I hope it helps:

- swap in replacement for `useMutation`
- exposes a `reset` callback

```ts
import {ApolloError, MutationResult, MutationTuple, useMutation} from '@apollo/client';
import {MutationHookOptions} from '@apollo/client/react/types/types';
import {DocumentNode} from 'graphql';
import {useState} from 'react';

interface IResettableMutationState<TData = any>
  extends Pick<MutationResult<TData>, 'data' | 'called' | 'loading' | 'error'> {}

export function useResettableMutation<TData = any, TVariables = any>(
  query: DocumentNode,
  props: MutationHookOptions<TData, TVariables>
): [MutationTuple<TData, TVariables>[0], MutationResult<TData> & {reset: () => void}] {
  const [{loading, data, error, called}, setState] = useState<IResettableMutationState>({
      called: false,
      loading: false
    }),
    onCompleted = (d: TData) => {
      setState({data: d, loading: false, called, error});
      props.onCompleted && props.onCompleted(data);
    },
    onError = (e: ApolloError) => {
      setState({loading: false, called, error: e});
      props.onError && props.onError(e);
    },
    reset = () => {
      setState({data: undefined, loading: false, called: false, error: undefined});
    };

  const [mutate, {client}] = useMutation<TData, TVariables>(query, {
      ...props,
      onCompleted,
      onError
    }),
    mutateWrapper = (opts: Parameters<MutationTuple<TData, TVariables>[0]>[0]) => {
      setState({data: undefined, loading: true, called: true, error: undefined});
      return mutate(opts);
    };

  return [mutateWrapper, {data, loading, error, called, client, reset}];
}
```

**Comment by bmitchinson** [comment]
> This feature would have been handy indeed. Implemented a quick wrapper to work around the limitation for now.
> Sharing below, I hope it helps:
> 
> * swap in replacement for `useMutation`
> * exposes a `reset` callback
> 
> ```ts
> import {ApolloError, MutationResult, MutationTuple, useMutation} from '@apollo/client';
...

Using this as a drop-in replacement for useMutation has been great @marc-ed-raffalli. Thanks so much for taking the time to leave it here. Will be sure to note any improvements we come across in usage.

**Comment by marc-ed-raffalli** [comment]
@markymc good idea. Packing that, may as well throw in a few utility hooks I wrote.

**Comment by markymc** [comment]
Nice! Please give me a shout when it’s up! Thanks!

**Comment by marc-ed-raffalli** [comment]
@markymc package published, please check it out and have a try:

https://www.npmjs.com/package/apollo-hooks-extended

**Comment by kevinrodriguez-io** [comment]
Any news/ideas?

**Comment by huykon** [comment]
any updated?

**Comment by MartinDevillers** [comment]
Ran into this limitation today. I have a form that's tied to a `useMutation` with some validation happening on the server. In case of validation errors, I unpack the `error` object and present it to the user in a dialog to let him/her know what went wrong. Initially, I wanted to clear the `error` object when the user dismisses the dialog, but since it came from the hook there was no way to mutate or reset it (as it should). Instead, I dropped the `error` variable and switched to the `onError` handler instead. This is a callback which uses the same `error` object. I then wire that to my own stateful variable. This gives me the exact same `error` variable *and* I can use `setError` to clear it. See example below:
```
  const [error, setError] = React.useState(null)
  const [uploadCoffee] = useMutation(UPLOAD_COFFEE, {
    onCompleted() {
      // ...
    },
    onError: setError,
  })
```

**Comment by kestutisg** [comment]
> Ran into this design quirk today. I have a form that's tied to a `useMutation` with some validation happening on the server. In case of validation errors, I unpack the `error` object and present it to the user in a dialog to let him/her know what went wrong. Initially, I wanted to clear the `error` object when the user dismisses the dialog, but since it came from the hook there was no way for me to mutate or reset it. Instead, I dropped the `error` variable from the hook and switched to the `onError` handler instead. This is a callback which uses the same `error` object. I then wire that to my own stateful variable. This gives me the exact same `error` variable _and_ I can use `setError` to clear it. See example below:
> 
> ```
>   const [error, setError] = React.useState(null)
>   const [uploadSomething] = useMutation(UPLOAD_SOMETHING, {
>     onCompleted() {
>       // ...
>     },
>     onError: setError,
>   })
> ```

@MartinDevillers and you don't run into https://github.com/apollographql/apollo-client/issues/5708 issue?

**Comment by MartinDevillers** [comment]
> @MartinDevillers and you don't run into [apollographql/apollo-client#5708](https://github.com/apollographql/apollo-client/issues/5708) issue?

No, but that may be that for my specific use-case I'm only interested in validation errors on the server side. These are returned by Apollo Server as a "soft-error", which is a regular HTTP 200 response with a JSON-payload like `{ data: null, errors: [...] }`. I'm not sure how other failures like network errors or non-200 responses are handled in my scenario (that's probably worth a test or two)

**Comment by freynolds** [comment]
I just reloaded the page after the error handler... not at all a good solution but works for me.

**Comment by marc-ed-raffalli** [comment]
@freynolds, have you checked https://www.npmjs.com/package/apollo-hooks-extended as a temporary solution ?

**Comment by marc-ed-raffalli** [comment]
@bordeo Since the comment I published a package: 
https://www.npmjs.com/package/apollo-hooks-extended
It has the fix for the error handling.

**Comment by MarcusOy** [comment]
Thank you @marc-ed-raffalli , I can't believe this isn't integrated into Apollo Client already...

**Comment by aquelehugo** [comment]
Does the Apollo team even read any of this? I guess no. You may have better luck opening an issue in the Apollo Client repo. :shrug:
